### ISSUE TITLE
Changes regarding JavaScript target output

### ISSUE THREAD

**Issue by Endercheif** [gleam-lang/gleam#2106]
There are 2 minor changes I'd like to contribute regarding JavaScript output. (Making one issue because they are so small).             

1. `export { ... } from "..."` rather than import then export

**Gleam**
```gleam
pub external fn add(x: Float, y: Float) -> Float = "./mod.js" "add"
pub external fn sub(x: Float, y: Float) -> Float = "./mod.js" "sub_func"
```

**Current JS**
```js
import { add, sub_func as sub } from "./mod.js";

export { add, sub };

```

**Preferred JS**
```js 
export { add, sub_func as sub } from "./mod.js";

```

2. Since data is immutable, variables outputted should be const.

**Gleam**
```gleam
pub fn main() {
  let x = 1
  let y = 6
  x + y
}
```

**Current JS**
```js
export function main() {
  let x = 1;
  let y = 6;
  return x + y;
}
```

**Preferred JS**
```js
export function main() {
  const x = 1;
  const y = 6;
  return x + y;
}
```
additionally *most* of [prelude.js](https://github.com/gleam-lang/gleam/blob/9097376f510411c57c0e8934a563723c4730e813/compiler-core/templates/prelude.js) can be `const` as well. this does provide slight performance improvements to the JS VM
Some places even use `var` which is problematic sometimes in its own right.

**Comment by lpil** [comment]
That change to the import/export look good!

I opted for `const` because I thought the name was misleading- it doesn't define a constant, the value can be freely mutated. If there's performance benefit that would be a good reason to use it. I did some searching but I couldn't find any information on this, do you have any?

**Comment by lpil** [comment]
I just gave this a test and `export { add, sub_func as sub } from "./mod.js";` doesn't allow the current module to use `add` and `sub`, only its dependants, so we wouldn't be able to do this always. We'd to save which functions are used in state somewhere and pass that to the codegen so it knows which format to use. I'm sure it is worthwhile, what do you think?

**Comment by Endercheif** [comment]
> That change to the import/export look good!
> 
> I opted not to use `const` because I thought the name was misleading- it doesn't define a constant, the value can be freely mutated. If there's performance benefit that would be a good reason to use it. I did some searching but I couldn't find any information on this, do you have any?

Sorry for the very very late reply. Forgot about Github notifications, went on vacation for a whole month and then technical difficulties meaning I thought I replied a few weeks ago but did not :p 

Regarding this, it must have just been something I misremembered because in testing the performance difference was minimal. `var` on the other hand is the weird one. Similar to functions, `hoisting` occurs to any variables declared with it, also if it is not in the function it is a global variable (can be accessed under `globalThis.FOO` / `window.FOO`. 

  

> I just gave this a test and `export { add, sub_func as sub } from "./mod.js";` doesn't allow the current module to use `add` and `sub`, only its dependants, so we wouldn't be able to do this always. We'd to save which functions are used in state somewhere and pass that to the codegen so it knows which format to use. I'm sure it is worthwhile, what do you think?

Ah, I was not aware of that, I guess since I rarely come across it in just JavaScript. I'm not really sure what the best course of action is. For Gleam in cases when things are just being re-exported without being used, The amount of lines is doubled. (ex, I have an auto generated `.gleam` file of functions that has 117 functions but when compiled to JavaScript 234 lines worth of export/import, when the `export { ... } from "..."` syntax would keep it at 117. But then again, it's only a matter of kilobytes so I'm not sure if it is even worth it. That double line file is only `3.96 kB` (`1.19 kB` compressed).

Long story short, I'd say for the effort to make it possible, it's not worth it for the minimal gain.

**Comment by lpil** [comment]
Thank you
